File: Linear_Control_Course/Chapter5/Lesson2/Chapter5_Lesson2.cpp
#include <iostream>
#include <vector>
#include <cmath>

int main() {
    double K   = 2.0;
    double tau = 0.5;
    double dt  = 0.001;
    double T   = 5.0;

    int N = static_cast<int>(T / dt);
    std::vector<double> t_vec(N), y_vec(N);

    double y = 0.0; // initial condition

    for (int k = 0; k < N; ++k) {
        double t = k * dt;

        // Choose test input:
        double u_step    = 1.0;           // unit step
        double u_ramp    = t;             // unit ramp
        double u_impulse = (k == 0) ? 1.0 / dt : 0.0; // discrete impulse

        double u = u_step; // switch to u_ramp or u_impulse for other tests

        // First-order ODE: tau dy/dt + y = K u
        double dydt = (-y + K * u) / tau;
        y += dt * dydt;

        t_vec[k] = t;
        y_vec[k] = y;
    }

    // Print a few samples
    for (int k = 0; k < N; k += N / 10) {
        std::cout << "t = " << t_vec[k]
                  << ", y = " << y_vec[k] << std::endl;
    }

    // In robotics, this loop can be placed inside a real-time control thread,
    // with u read from sensor-based trajectories or ROS topics.
    return 0;
}

File: Linear_Control_Course/Chapter5/Lesson2/Chapter5_Lesson2.py
import numpy as np
import matplotlib.pyplot as plt

# Parameters of first-order system: G(s) = K / (tau s + 1)
K = 2.0
tau = 0.5

# Time grid
t_end = 5.0
n_points = 2000
t = np.linspace(0.0, t_end, n_points)

# Analytical responses for unit inputs (U0 = 1, R = 1)
u_step = np.ones_like(t)          # unit step
u_ramp = t                        # unit ramp
# Approximate unit impulse as a narrow pulse of width dt and area 1
dt = t[1] - t[0]
u_imp = np.zeros_like(t)
u_imp[0] = 1.0 / dt

# Closed-form responses
y_step = K * (1.0 - np.exp(-t / tau))           # y_step(t)
y_ramp = K * (t - tau * (1.0 - np.exp(-t / tau)))  # y_ramp(t)
y_imp = (K / tau) * np.exp(-t / tau)            # y_imp(t)

# (Optional) numerical integration of the ODE tau dy/dt + y = K u(t)
def simulate_first_order(u):
    y = np.zeros_like(t)
    for k in range(1, len(t)):
        dydt = (-y[k-1] + K * u[k-1]) / tau
        y[k] = y[k-1] + dt * dydt
    return y

y_step_num = simulate_first_order(u_step)
y_ramp_num = simulate_first_order(u_ramp)
y_imp_num = simulate_first_order(u_imp)

# Plotting
plt.figure()
plt.plot(t, y_step, label="analytic step")
plt.plot(t, y_step_num, "--", label="numerical step")
plt.xlabel("t [s]")
plt.ylabel("y_step(t)")
plt.legend()

plt.figure()
plt.plot(t, y_ramp, label="analytic ramp")
plt.plot(t, y_ramp_num, "--", label="numerical ramp")
plt.xlabel("t [s]")
plt.ylabel("y_ramp(t)")
plt.legend()

plt.figure()
plt.plot(t, y_imp, label="analytic impulse")
plt.plot(t, y_imp_num, "--", label="numerical impulse")
plt.xlabel("t [s]")
plt.ylabel("y_imp(t)")
plt.legend()

plt.show()

# In a robotics context:
# - G(s) may model a motor velocity loop.
# - Libraries such as 'control' (python-control) can define tf(K, [tau, 1])
#   and compute step_response/impulse_response for joint-level design.
